test/prober/prober.go

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2020 The Knative Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package prober
+
+import (
+	"go.uber.org/zap"
+	"knative.dev/eventing/test/lib"
+	"testing"
+	"time"
+)
+
+var (
+	Version = "v0.7.0"
+	// FIXME: Interval is set to 200 msec, as lower values will result in errors
+	// https://github.com/knative/eventing/issues/2357
+	// Interval = 10 * time.Millisecond
+	Interval = 200 * time.Millisecond
+)
+
+// Prober is the interface for a prober, which checks the result of the probes when stopped.
+type Prober interface {
+	// Verify will verify prober state after finished has been send
+	Verify() ([]error, int)
+
+	// Finish send finished event
+	Finish()
+
+	// deploy a prober to a cluster
+	deploy()
+	// remove a prober from cluster
+	remove()
+}
+
+// Config represents a configuration for prober
+type Config struct {
+	Namespace     string
+	Interval      time.Duration
+	Serving       ServingConfig
+	FinishedSleep time.Duration
+}
+
+type ServingConfig struct {
+	Use         bool
+	ScaleToZero bool
+}
+
+func NewConfig(namespace string) *Config {
+	return &Config{
+		Namespace:     namespace,
+		Interval:      Interval,
+		FinishedSleep: 5 * time.Second,
+		Serving: ServingConfig{
+			Use:         true,
+			ScaleToZero: true,
+		},
+	}
+}
+
+// RunEventProber starts a single Prober of the given domain.
+func RunEventProber(log *zap.SugaredLogger, client *lib.Client, config *Config) Prober {
+	pm := newProber(log, client, config)
+	pm.deploy()
+	return pm
+}
+
+// AssertEventProber will send finish event and then verify if all events propagated well
+func AssertEventProber(t *testing.T, prober Prober) {
+	prober.Finish()
+
+	waitAfterFinished(prober)
+
+	errors, events := prober.Verify()
+	if len(errors) == 0 {
+		t.Logf("All %d events propagated well", events)
+	} else {
+		t.Logf("There ware %v errors. Listing them below.", len(errors))
+	}
+	for _, err := range errors {
+		t.Error(err)
+	}
+
+	prober.remove()
+}
+
+type prober struct {
```

Comment:
This type seems to have sooo many functions attached to it. I'm wondering if they could be split into multiple objects or packages. I understand all of them are there to achieve one goal but they could be split into cohesive packages/objects.

Reply:
Right. I'll see how to split it to more manageable chunks. I'd like to keep it under one package, but subpackages are a possibility.

Reply:
I think I reduce the code on prober to minimum. There are left only things that directly related to deploy and removal of wathola components and verify of results.